File: Linked_List_Cycle.cpp
#include "leetcode.hpp"

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(nullptr) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode *follower = head;
        while (head != nullptr && head->next != nullptr) {
            follower = follower->next;
            head = head->next->next;

            if (follower == head)
                return true;
        }
        return false;
    }
};

File: Linked_List_Cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:

    # @param head, a ListNode
    # @return a boolean
    def hasCycle(self, head):
        follower = head
        while head is not None and head.next is not None:
            follower = follower.next
            head = head.next.next
            if follower == head:
                return True
        return False
